### PR TITLE
N8n 4040 data mapping

### PR DESCRIPTION
**Pull request by mutdmour**

### CODE REVIEW

**Review by MiloradFilipovic — APPROVED** [review]
Nice job! 🙌 Really looking forward to this feature.
I found a few nitpicks and possible refactor possibilities so once that is discussed/addressed this is good to go as far as I am concerned.

Some more stuff unrelated to code:
- Personally, I expect from dragged value to overwrite current expression, not append to it so I find it confusing and a bit annoying that I have to go to editor and remove previous values. This is especially confusing since most of the time, only a single value (data pill) is visible in input field.
- Just found out that `combine` field in the `IF` node accepts expressions. Haven't found anything regarding possible values other that `ALL` and `ANY`. Is this a bug?
if_node_combine

File: packages/editor-ui/src/components/InputPanel.vue [hunk 1]
```diff
@@ -182,6 +213,26 @@ export default mixins(
...
+					if (this.draggableHintShown) {
+						return;
+					}
+					this.showDraggableHintWithDelay = this.showDraggableHint;
+					if (this.showDraggableHintWithDelay) {
+						this.draggableHintShown = true;
+
+						this.$telemetry.track('User viewed data mapping tooltip', { type: 'unexecuted input pane' });
+					}
+				}, 1000);
```
[the remark this answers is not in the document]
Comment (mutdmour) [reply packages/editor-ui/src/components/InputPanel.vue:229]: with timeouts like this I don't think it's as important..

File: packages/editor-ui/src/components/ParameterInputExpanded.vue [hunk 2]
```diff
@@ -4,38 +4,53 @@
...
+				:parameter="parameter"
+				:value="value"
+				:isReadOnly="false"
+				:showOptions="true"
+				@optionSelected="optionSelected"
+				@menu-expanded="onMenuExpanded"
+			/>
+		</template>
+		<template>
+			<parameter-input
```
[the remark this answers is not in the document]
Comment (mutdmour) [reply packages/editor-ui/src/components/ParameterInputExpanded.vue:20]: too complicated to move rn... That was the goal.. Idea was as design changed we would start to use Expanded everywhere... that's the one used in credentials right now. though does not seem like the designs will converge... so maybe we should rename one to ParamterInputCredentials and other to ParameterInputNDV

File: packages/editor-ui/src/components/RunDataTable.vue [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+					src_field_name: column,
+					src_nodes_back: this.distanceFromActive,
+					src_run_index: this.runIndex,
+					src_runs_total: this.totalRuns,
+					src_view: 'table',
+					src_element: 'column',
+					success: false,
+					...mappingTelemetry,
+				});
+			}, 1000); // ensure dest data gets set if drop
```
[the remark this answers is not in the document]
Comment (mutdmour) [reply packages/editor-ui/src/components/RunDataTable.vue:165]: this number here is meaningless..

File: packages/editor-ui/src/components/RunDataTable.vue [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+		focusedMappableInput(): string {
+			return this.$store.getters['ui/focusedMappableInput'];
+		},
+		showHint (): boolean {
+			return !this.draggedColumn && (this.showMappingHint || (!!this.focusedMappableInput && window.localStorage.getItem(LOCAL_STORAGE_MAPPING_FLAG) !== 'true'));
+		},
+	},
+	methods: {
+		shorten(s: string) {
+			return shorten(s, 16, 2);
```
[the remark this answers is not in the document]
Comment (mutdmour) [reply packages/editor-ui/src/components/RunDataTable.vue:121]: I don't think so

[the remark this answers is not in the document]
Comment (mutdmour) [reply packages/editor-ui/src/components/RunDataTable.vue:121]: actually can add it a local method and use it directly in template
```suggestion
			shorten,
```

**Comment by mutdmour** [comment]
@MiloradFilipovic 

> Personally, I expect from dragged value to overwrite current expression, not append to it so I find it confusing and a bit annoying that I have to go to editor and remove previous values. This is especially confusing since most of the time, only a single value (data pill) is visible in input field.

I agree with you.. discussed this a lot but Max is confident in this.

> Just found out that combine field in the IF node accepts expressions. Haven't found anything regarding possible values other that ALL and ANY. Is this a bug?

not a bug.. can be disabled directly in if node definition if it's not supposed to

File: packages/editor-ui/src/components/ParameterInputExpanded.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (MiloradFilipovic) [reply packages/editor-ui/src/components/ParameterInputExpanded.vue:20]: Maybe, that names better indicate where the components are used but still make you think it's different variations of the same component. We can leave it as is rn.

**Comment by janober** [comment]
Amazing! Just got merged.

**Comment by janober** [comment]
Got released with `n8n@0.187.1`

**Pull request merged**